server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -261,10 +261,12 @@ private static LocalCheckpointTracker createLocalCheckpointTracker(EngineConfig
             // Thus, we need to restore the LocalCheckpointTracker bit by bit to ensure the consistency between LocalCheckpointTracker and
             // Lucene index. This is not the only solution since we can bootstrap max_seq_no_of_updates with max_seq_no of the commit to
             // disable the MSU optimization during recovery. Here we prefer to maintain the consistency of LocalCheckpointTracker.
-            if (localCheckpoint < maxSeqNo && engineConfig.getIndexSettings().isSoftDeleteEnabled()) {
+            // The max_seq_no of Lucene commit in the old indices might be smaller than seq_no of some documents in the commit.
+            // We have to rebuild the LocalCheckpointTracker for those indices. See https://github.com/elastic/elasticsearch/pull/38879.
```

Comment:
perhaps add a comment that this is only necessary for 6.6.0 and 6.6.1 indices (earlier indices did not have soft-deletes)